### DISCUSSION TITLE
Prebuilt Actions Import action

### DISCUSSION THREAD

**Discussion by kombiahrk** [filamentphp/filament#10150] [Help]
How to create relationship record if it is not available ?

i have both BelongsTo(Brand) , BelongsToMany(categories)  relationship type in Product Model and i am using ProductImporter to import

```
<?php

namespace App\Filament\Imports;

use Carbon\Carbon;
use App\Models\Product;
use Illuminate\Validation\Rule;
use Filament\Actions\Imports\Importer;
use Filament\Forms\Components\Checkbox;
use Filament\Actions\Imports\ImportColumn;
use Filament\Actions\Imports\Models\Import;

class ProductImporter extends Importer
{
    protected static ?string $model = Product::class;

    public static function getColumns(): array
    {
        return [
            ImportColumn::make('name')
                ->label('Name')
                ->requiredMapping()
                ->rules(['required', 'max:255']),

            ImportColumn::make('barcode')
                ->label('Barcode')
                ->requiredMapping()
                ->rules([
                    'required',
                    'max:255',
                    Rule::unique('products', 'barcode')
                        ->where(function ($query) {
                            return $query->where('team_id', filament()->getTenant()->id);
                        })
                ]),

            ImportColumn::make('hsn_code')
                ->label('HSN Code')
                ->rules(['max:255']),

            ImportColumn::make('tax')
                ->label('Tax')
                ->relationship(resolveUsing: ['name', 'rate']),

            ImportColumn::make('stock_alert')
                ->label('Stock Alert')
                ->ignoreBlankState(),
            ImportColumn::make('quantity_limit')
                ->label('Quantity Limit')
                ->ignoreBlankState(),

            ImportColumn::make('brand')
                ->label('Brand')
                ->relationship(resolveUsing: ['name', 'slug']),

            ImportColumn::make('mrp_price')
                ->label('MRP Price')
                ->castStateUsing(function (?string  $state) {
                    if (blank($state)) {
                        return 0;
                    }
                })
                ->numeric(),
            ImportColumn::make('pur_price')
                ->label('Purchase Price(Tax-Exclusive)')
                ->castStateUsing(function (?string  $state) {
                    if (blank($state)) {
                        return 0;
                    }
                })
                ->numeric(),
            ImportColumn::make('sell_price')
                ->label('Selling Price(Tax-Exclusive)')
                ->castStateUsing(function (?string  $state) {
                    if (blank($state)) {
                        return 0;
                    }
                })
                ->numeric(),

            ImportColumn::make('quantity')
                ->label('Current Quantity')
                ->castStateUsing(function (?string  $state) {
                    if (blank($state)) {
                        return 0;
                    }
                })
                ->numeric(),
            ImportColumn::make('as_on_date')
                ->label('As on date')
                ->castStateUsing(function (?string  $state) {
                    if (blank($state)) {
                        return today();
                    }
                    // Convert to Carbon instance
                    $carbon_date = Carbon::createFromFormat('d-m-Y', $state);

                    // Format the date as yyyy/mm/dd
                    $formatted_date = $carbon_date->format('Y-m-d');

                    return $formatted_date;
                })
                ->rules(['required']),


            ImportColumn::make('unit')
                ->label('Base Unit')
                ->requiredMapping()
                ->relationship(resolveUsing: ['name', 'short_name']),

            // ImportColumn::make('enable_alt_units')
            //     ->label('Secoundary Unit')
            //     ->requiredMapping()
            //     ->boolean()
            //     ->rules(['required', 'boolean']),

            // ImportColumn::make('alternate_units.alt_unit_id')
            //     ->label('Alternate Unit')
            //     ->relationship(resolveUsing: ['name', 'short_name']),
            // ImportColumn::make('alternate_units.alt_unit_value')
            //     ->label('Conversion Value'),

            ImportColumn::make('status')
                ->label('Active')
                ->requiredMapping()
                ->boolean()
                ->rules(['required', 'boolean']),
        ];
    }

    public function resolveRecord(): ?Product
    {
        // if ($this->options['update_existing'] ?? false) {
        //     return Product::firstOrNew([
        //         // Update existing records, matching them by `$this->data['column_name']`
        //         //'team_id' => filament()->getTenant()->id,
        //         //'barcode' => $this->data['barcode']
        //     ]);
        // }
        $newProduct = new Product();
        $newProduct->team_id = filament()->getTenant()->id;
        return $newProduct;
    }

    public static function getCompletedNotificationBody(Import $import): string
    {
        $body = 'Your product import has completed and ' . number_format($import->successful_rows) . ' ' . str('row')->plural($import->successful_rows) . ' imported.';

        if ($failedRowsCount = $import->getFailedRowsCount()) {
            $body .= ' ' . number_format($failedRowsCount) . ' ' . str('row')->plural($failedRowsCount) . ' failed to import.';
        }

        return $body;
    }

    // public static function getOptionsFormComponents(): array
    // {
    //     return [
    //         Checkbox::make('update_existing')
    //             ->label('Update existing records'),
    //     ];
    // }
}
```

**Comment by MasterAdmin2** [comment]
let's understand using country state city example,
as country has many states and states has many cities and also belongs to relationship,

I approached by making this code work,

`// Retrieve the latest file path from the imports table
    $latestFilePath = DB ::table ( 'imports' )
        -> latest ()
        -> value ( 'file_path' );

    // Access the file contents using the full path
    $fileContents = File ::get ( $latestFilePath );

    // Create a CSV reader instance from the file contents
    $csv = Reader ::createFromString ( $fileContents );

    // Iterate through each city record
    foreach ( $csv -> getRecords () as $record ) {
        $cityData = array_values ( $record ); // Adjust for your column order
        $city = $cityData[ 0 ];
        $state = $cityData[ 1 ];
        $country = $cityData[ 2 ];

        // Find or create the country
        $country = Country ::firstOrCreate ( [ 'name' => $country ] );

        // Find or create the state
        $state = State ::firstOrCreate ( [ 'name' => $state , 'country_id' => $country -> id ] );

        // Create the city record
        $createdCityNames = City ::create ( [
            'name' => $city ,
            'state_id' => $state -> id ,
            'country_id' => $country -> id ,
        ] );
    }`

now I wanted to understand this how can I configure this piece of code with my importer,
adding importer code as well

`<?php

    namespace App\Filament\Imports;

    use App\Models\City;
    use Filament\Actions\Imports\ImportColumn;
    use Filament\Actions\Imports\Importer;
    use Filament\Actions\Imports\Models\Import;
    use Filament\Forms\Components\Checkbox;
    use Illuminate\Support\Str;

    class CityImporter extends Importer
    {
        protected static ?string $model = City::class;

        public static function getColumns () : array
        {
            return [
                ImportColumn ::make ( 'name' )
                    -> requiredMapping ()
                    -> fillRecordUsing ( fn ( City $record , string $state ) => $record -> name = Str ::title ( $state ) )
                    -> rules ( [ 'required' , 'max:191' ] ) ,
                ImportColumn ::make ( 'state' )
                    -> requiredMapping ()
                    -> relationship ()
                    -> rules ( [ 'required' ] ) ,
                ImportColumn ::make ( 'country' )
                    -> requiredMapping ()
                    -> relationship ()
                    -> rules ( [ 'required' ] ) ,
            ];
        }

        public function resolveRecord () : ?City
        {
            if ( $this -> options[ 'updateExisting' ] ?? false ) {
                return City ::firstOrNew ( [
                    // Update existing records, matching them by `$this->data['column_name']`
                    'name' => $this -> data[ 'name' ] ,
                ] );
            }
            return new City();
        }

        public static function getCompletedNotificationBody ( Import $import ) : string
        {
            $body = 'Your city import has completed and ' . number_format ( $import -> successful_rows ) . ' ' . str ( 'row' ) -> plural ( $import -> successful_rows ) . ' imported.';

            if ( $failedRowsCount = $import -> getFailedRowsCount () ) {
                $body .= ' ' . number_format ( $failedRowsCount ) . ' ' . str ( 'row' ) -> plural ( $failedRowsCount ) . ' failed to import.';
            }

            return $body;
        }


        public static function getOptionsFormComponents () : array
        {
            return [
                Checkbox ::make ( 'updateExisting' )
                    -> label ( 'Update existing records' ) ,
            ];
        }
    }
`

can you help me if possible.?
